File: 백준/Silver/14501. 퇴사/퇴사.py
import sys
input=lambda:sys.stdin.readline().rstrip()
n=int(input())
ti,pi=[],[]
for _ in range(n):
    t,p=map(int,input().split())
    ti+=[t]
    pi+=[p]
dp=[0]*(n+1)
def solve(day):
    if n<=day: return 0
    if dp[day]: return dp[day]
    case1=solve(day+1)
    case2=0
    if day+ti[day]<=n:
        case2=solve(day+ti[day])+pi[day]
    dp[day]=max(case1,case2)
    return dp[day]
print(solve(0))

File: 백준/Silver/14501. 퇴사/퇴사.cc
#include <iostream>
#include <vector>
using namespace std;
int solve(vector<int> &ti, vector<int> &pi, vector<int> &memo, int n, int day) {
    int case1, case2;
    if (n <= day) return 0;
    if (memo[day]) return memo[day];
    case1 = solve(ti, pi, memo, n, day + 1), case2 = 0;
    if (day + ti[day] <= n)
        case2 = solve(ti, pi, memo, n, day + ti[day]) + pi[day];
    memo[day] = max(case1, case2);
    return memo[day];
}
int main(void) {
    int n;
    cin >> n;
    vector<int> ti(n), pi(n), memo(n + 1, 0);
    for (int i = 0; i < n; i++)
        cin >> ti[i] >> pi[i];
    cout << solve(ti, pi, memo, n, 0);
    return 0;
}